Added async audit logger which uses Apache Derby for temp storage

ssrangisetti opened the pull request.

saisandeep-flipkart committed:
Added a very simple async audit logger which just writes audits in separate virtual thread

saisandeep-flipkart committed:
Adding derby based async audit logging

saisandeep-flipkart committed:
Added unit tests

ssrangisetti commented:
@coderabbitai review

saisandeep-flipkart committed:
Fixed a bug where timestamp is coming as null in AuditEntry

ssrangisetti reviewed (commented):
We need to put limits on user given data: username and password lengths

ssrangisetti commented on derby-async-audit/src/main/java/com/flipkart/grayskull/app/audit/DerbyAsyncAuditLogger.java:
```diff
@@ -0,0 +1,95 @@
+            while (resultSet.next()) {
+                maxId = resultSet.getLong(1);
+                String eventSting = resultSet.getString(2);
+                AuditEntry auditEntry = objectMapper.readValue(eventSting, AuditEntry.class);
```
we can add unique id of logs which is nodename + id to identify duplicate log entries

ssrangisetti commented on server/src/main/java/com/flipkart/grayskull/controllers/SecretController.java:
```diff
@@ -57,6 +66,10 @@ public ResponseTemplate<SecretMetadata> readSecretMetadata(@PathVariable("projec
     public ResponseTemplate<SecretDataResponse> readSecretValue(@PathVariable("projectId") @NotBlank @Size(max = 255) String projectId, @PathVariable("secretName") @NotBlank @Size(max = 255) String secretName) {
         SecretDataResponse response = secretService.readSecretValue(projectId, secretName);
+        Map<String, String> auditMetadata = new HashMap<>();
+        SanitizingObjectMapper.addToMap(auditMetadata, "result", response);
```
add IP address in the audits

ssrangisetti replied:
only "dataVersion" is sufficient in audits and "publicPart" is good to have to audit

ssrangisetti commented on server/src/main/java/com/flipkart/grayskull/controllers/SecretController.java:
```diff
@@ -57,6 +66,10 @@ public ResponseTemplate<SecretMetadata> readSecretMetadata(@PathVariable("projec
         SecretDataResponse response = secretService.readSecretValue(projectId, secretName);
+        Map<String, String> auditMetadata = new HashMap<>();
+        SanitizingObjectMapper.addToMap(auditMetadata, "result", response);
+        String userId = SecurityContextHolder.getContext().getAuthentication().getName();
```
we need the audit on super-admin read api as well.

ssrangisetti commented on server/src/main/java/com/flipkart/grayskull/audit/utils/SanitizingObjectMapper.java:
```diff
@@ -27,4 +34,15 @@ public static ObjectMapper create() {
+        try {
+            map.put(key, MASK_OBJECT_MAPPER.writeValueAsString(value));
+        } catch (JsonProcessingException e) {
+            map.put(key, "Error serializing object: " + e.getMessage());
```
we should not be ignoring error. but we should pass it to the user

saisandeep-flipkart committed:
Addressed review comments

saisandeep-flipkart committed:
Merge branch 'mvp-rest-apis' of https://github.com/imanadi/grayskull into derby-audit

# Conflicts:
#	pom.xml
#	server/src/main/java/com/flipkart/grayskull/audit/AuditAction.java
#	server/src/main/java/com/flipkart/grayskull/audit/AuditAspect.java
#	server/src/main/java/com/flipkart/grayskull/controllers/SecretController.java
#	spi/src/main/java/com/flipkart/grayskull/spi/models/AuditEntry.java

saisandeep-flipkart committed:
Merge branch 'mvp-rest-apis' of https://github.com/imanadi/grayskull into derby-audit

saisandeep-flipkart committed:
fixed the bug where audit logs were not getting persisted

 audit logs were not getting persisted because of class cast exception. so create a mapper which converts audit entry to entity for persisting

saisandeep-flipkart committed:
Adding created, last modified times for AuditCheckpoint

grssam commented on derby-async-audit/src/main/java/com/flipkart/grayskull/app/audit/DerbyAsyncAuditLogger.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    @PostConstruct
+    public void init() throws SQLException {
```
as discussed - we need to make sure that any stale derby db on the disk (in case of persistent volume being reattached to a compute node) is cleaned up first based on some staleness config.

grssam commented on derby-async-audit/src/main/java/com/flipkart/grayskull/app/audit/DerbyAsyncAuditLogger.java:
```diff
@@ -0,0 +1,96 @@
+
+    @Override
+    public void log(AuditEntry auditEntry) {
+        try (PreparedStatement statement = connection.prepareStatement("INSERT INTO audits (event) VALUES (?)")) {
```
what's the point of preparing a statement and using it only once? prepared statements are supposed to be reusable..

ssrangisetti replied:
I found out that derby actually as statement caching which caches all statements including prepared ones. https://db.apache.org/derby/docs/10.5/tuning/ctundepth32379.html. I have verified that prepared statements are indeed getting cached and reused

grssam commented on derby-async-audit/src/main/java/com/flipkart/grayskull/app/audit/DerbyAsyncAuditLogger.java:
```diff
@@ -0,0 +1,96 @@
+                auditEntries.add(auditEntry);
+            }
+            if (!auditEntries.isEmpty()) {
+                auditEntryRepository.saveAll(auditEntries);
```
can you add some logs in this method?

saisandeep-flipkart committed:
addressing review comments

grssam commented on derby-async-audit/src/main/java/com/flipkart/grayskull/app/audit/AuditProperties.java:
```diff
@@ -41,4 +43,10 @@ public class AuditProperties {
+     * If the checkpoint is older than this duration, then the data present in the Derby db is considered and deleted.
+     * If the duration is set to zero, then staleness check is disabled
+     */
+    private Duration stalenessDuration = Duration.ZERO;
```
should the default be some actual realistic value? else, the default here doesn't make any sense because it will always delete the existing DB.. (as the last checkpoint time would always be more than 0 ms/s ago)

ssrangisetti replied:
staleness check is disabled when duration <= 0. so by default we will not be deleting any data.

grssam commented on derby-async-audit/src/main/java/com/flipkart/grayskull/app/audit/AuditProperties.java:
```diff
@@ -41,4 +43,10 @@ public class AuditProperties {
     private String batchTimeInterval;
 
+    /**
+     * If the checkpoint is older than this duration, then the data present in the Derby db is considered and deleted.
```
considered "stale"

grssam commented on derby-async-audit/src/main/java/com/flipkart/grayskull/app/audit/DerbyAsyncAuditLogger.java:
```diff
@@ -30,7 +30,7 @@ public class DerbyAsyncAuditLogger implements AsyncAuditLogger {
 
     public DerbyAsyncAuditLogger(AuditProperties auditProperties, ObjectMapper objectMapper, MeterRegistry meterRegistry, AuditEntryRepository auditEntryRepository, AuditCheckpointRepository auditCheckpointRepository) throws SQLException {
-        this.connection = DriverManager.getConnection(auditProperties.getDerbyUrl());
+        this.connection = DriverManager.getConnection("jdbc:derby:" + auditProperties.getDerbyDirectory() + ";create=true");
```
this is being generated at a couple of places.. maybe there should be a common method to do this.. maybe even a derived auditproperties class getter which isn't part of actual config, but uses the derby directory to create the derby jdbc url

saisandeep-flipkart committed:
fixed a typo and added a derived getter for derby url

saisandeep-flipkart committed:
Adding lombok builder.default on timestamp fields

saisandeep-flipkart committed:
Merge branch 'mvp-rest-apis' of https://github.com/imanadi/grayskull into derby-audit

saisandeep-flipkart committed:
making derby audit logger primary

imanadi committed:
Feat: Initial Implementation of Grayskull Secret Management Service (#3)

### Overview
This PR establishes the foundational MVP implementation for Grayskull, a
secret management service. It introduces a modular, extensible
architecture with Spring Boot-based REST APIs, persistence-agnostic SPI
interfaces, MongoDB-backed storage, fine-grained authorization,
comprehensive auditing, and containerized local development setup.

---

### REST API Layer (`SecretController`)

**Endpoints:**
```
GET    /v1/projects/{projectId}/secrets
       - List secrets with pagination (offset/limit)
       - Secured: secrets.list permission
       - Returns latest version of each secret

POST   /v1/projects/{projectId}/secrets
       - Create new secret with initial data (version 1)
       - Secured: secrets.create permission
       - Supports SELF provider with public/private parts

GET    /v1/projects/{projectId}/secrets/{secretName}
       - Read secret metadata (name, version, timestamps, state)
       - Secured: secrets.read.metadata permission
       - Excludes sensitive data values

GET    /v1/projects/{projectId}/secrets/{secretName}/data
       - Read decrypted secret value
       - Secured: secrets.read.value permission
       - Returns both public and private parts

POST   /v1/projects/{projectId}/secrets/{secretName}/data
       - Upgrade secret data (creates new version)
       - Secured: secrets.update permission
       - Immutable versioning: previous versions preserved

DELETE /v1/projects/{projectId}/secrets/{secretName}
       - Soft-delete (marks as DISABLED)
       - Secured: secrets.delete permission
       - Audit trail maintained

GET    /v1/projects/{projectId}/secrets/{secretName}/versions/{version}
       - Admin API: Retrieve specific version
       - Secured: secrets.read.version permission
       - Supports state filtering (ACTIVE/DISABLED/DELETED)
```

**Request/Response Handling:**
- Jakarta Bean Validation for input sanitization (`@Valid`, `@NotBlank`,
`@Size`, constraint ranges)
- `ResponseTemplate<T>` wrapper with success/error indicators and
messages

---

### Architecture & Module Structure

#### 1. SPI Module (Service Provider Interface)
**Purpose:** Defines persistence-agnostic contracts and domain models
for extensibility.

**Key Components:**
- **Interfaces:**
- `GrayskullAuthorizationProvider`: Pluggable authorization provider SPI
- `SecretRepository`, `SecretDataRepository`, `ProjectRepository`,
`AuditEntryRepository`: Persistence-agnostic repository contracts
  
- **Domain Models:** `Secret`, `SecretData`, `Project`, `AuditEntry`
with lifecycle states (`ACTIVE`, `DISABLED`, `DELETED`)

- **Authorization Context:** `AuthorizationContext` for project and
secret-level permission checks

**Design Philosophy:** Zero external dependencies (only provided-scope
framework deps), enabling future implementations (e.g., different
databases, cloud provider integrations)

#### 2. Server Module
**Core Spring Boot application providing business logic and
persistence.**

##### Authorization Framework
**Implementation:**
- **`GrayskullSecurity`**: Security facade bean for `@PreAuthorize`
expressions
- `hasPermission(projectId, action)` - Project-level checks (e.g.,
create, list)
- `hasPermission(projectId, secretName, action)` - Secret-level checks
(e.g., read, update)
  
- **`SimpleAuthorizationProvider`**: Configuration-driven rule engine
with wildcard support
- Rules defined in `application.yml` under
`grayskull.authorization.rules`
  - Supports user, project, and secret-level matching with `*` wildcards
  
- **`AuthorizationProperties`**: Type-safe configuration binding for
authorization rules

**Action Definitions (`GrayskullActions`):** 
- `secrets.create`, `secrets.list`, `secrets.read.metadata`,
`secrets.read.value`, `secrets.read.version`, `secrets.update`,
`secrets.delete`

**Authorization Logic:**
- Non-existent projects → Transient project with default KMS key for
create operations
- Non-existent secrets → Fallback to project-level rules (allows
permission checks before resource exists)
- Existing resources → Full secret-level permission evaluation

##### Audit Framework (AOP-based)
**Implementation:**
- **`@Audit`** annotation with `AuditAction` enum for marking auditable
operations
- **`AuditAspect`**: Spring AOP aspect with `@AfterReturning` advice
  
**Behavior:**
- Audits only successful CUD (Create, Update, Delete) operations
- Captures: `userId`, `projectId`, `resourceType`, `resourceName`,
`resourceVersion`, `action`, `timestamp`
- Metadata includes sanitized request/response payloads
- **Masking:** `@AuditMask` annotation + custom Jackson serializer for
PII/sensitive data
- Transaction-aligned: Failures in audit serialization rollback the
entire operation

##### Service Layer
**`SecretService` / `SecretServiceImpl`:**
- Secret lifecycle management with versioning (immutable secret data)
- Soft-delete implementation (marks secrets as `DISABLED`)
- KMS key inheritance from project configuration
- Concurrency: Optimistic locking via MongoDB version field

##### Exception Handling
**`GlobalExceptionHandler`** (`@RestControllerAdvice`):
- Centralized exception handling with consistent error responses
- Validation errors return detailed `Violation` objects
- Custom exceptions mapped to appropriate HTTP status codes

##### Persistence Layer
**MongoDB Implementation:**
- Entity classes: `SecretEntity`, `SecretDataEntity`, `ProjectEntity`,
`AuditEntryEntity`
- Repository implementations: Delegates to `MongoRepository` interfaces
- Timestamp generation delegated to DB (`@CreatedDate`,
`@LastModifiedDate`)
- Separate collections for metadata and data (optimized reads)

##### Configuration
- **`CryptoConfig`**: Manages encryption keys (bound to
`grayskull.crypto.keys.*`)
- **`KmsConfig`**: Default KMS key configuration
- **`SecurityConfiguration`**: Spring Security with Basic Auth
- **`MongoBeanConfiguration`**: Auditing enablement, custom converters

#### 3. Setup Module
**Local Development Environment:**
- Docker Compose setup for 3-node MongoDB replica set (ports
37017-37019)
- Authentication with keyfile-based security
- Scripts: `setup.sh` (automated initialization), `teardown.sh`
(cleanup)
- Environment configuration via `.env` file

#### 4. Simple-App Module
**Purpose:** Example application demonstrating Grayskull integration as
a library.

---

### Testing Strategy

**Framework:** JUnit 5 + Spring Boot Test + Testcontainers

**`BaseIntegrationTest`:**
- Spins up ephemeral MongoDB container per test run
- `@DynamicPropertySource` for runtime configuration
- Provides helper methods for API operations with user context
- 12 test classes covering controllers, mappers, models, and
authorization

**Test Coverage:**
- **API Tests:** `SecretControllerIT` - CRUD operations, pagination,
versioning, authorization scenarios
- **Unit Tests:** `SecretMapperTest`, `SecretTest`, `SecretDataTest`,
`AuditMaskingSerializerTest`
- **Authorization Tests:** Multi-role scenarios (admin, editor, viewer)
with `@PreAuthorize` validation

**Test Configuration (`test/resources/application.yml`):**
- Predefined authorization rules for consistent test behavior
- User roles: `admin` (full access), `editor` (create/read/update),
`viewer` (read-only)

---

### Build & Tooling

**Maven Multi-Module:**
- Parent POM with centralized dependency management
- Spring Boot 3.5.0, Java 21
- Dependencies: Spring Data MongoDB, Spring Security, MapStruct, Lombok,
Jackson, Testcontainers

saisandeep-flipkart committed:
Merge branch 'main' of github.com:flipkart-incubator/grayskull into derby-audit

# Conflicts:
#	server/pom.xml
#	server/src/main/java/com/flipkart/grayskull/audit/AuditAction.java
#	server/src/main/java/com/flipkart/grayskull/audit/AuditAspect.java
#	server/src/main/java/com/flipkart/grayskull/audit/AuditConstants.java
#	server/src/main/java/com/flipkart/grayskull/audit/utils/SanitizingObjectMapper.java
#	server/src/main/java/com/flipkart/grayskull/controllers/SecretController.java
#	server/src/main/java/com/flipkart/grayskull/entities/AuditEntryEntity.java
#	server/src/main/java/com/flipkart/grayskull/mappers/SecretMapper.java
#	server/src/main/java/com/flipkart/grayskull/service/impl/SecretServiceImpl.java
#	server/src/main/java/com/flipkart/grayskull/service/interfaces/SecretService.java
#	server/src/main/java/com/flipkart/grayskull/spimpl/repositories/AuditEntryRepositoryImpl.java
#	server/src/test/java/com/flipkart/grayskull/BaseIntegrationTest.java
#	server/src/test/java/com/flipkart/grayskull/audit/utils/AuditMaskingSerializerTest.java
#	server/src/test/java/com/flipkart/grayskull/controllers/SecretControllerIT.java
#	server/src/test/java/com/flipkart/grayskull/mappers/SecretMapperTest.java
#	simple-app/src/main/resources/application.yml
#	simple-app/src/test/resources/application.yml
#	spi/src/main/java/com/flipkart/grayskull/spi/models/AuditEntry.java
#	spi/src/main/java/com/flipkart/grayskull/spi/repositories/AuditEntryRepository.java

grssam commented:
can you now raise a fresh PR against main repo?

saisandeep-flipkart committed:
addressed AI code review comments
